zap/src/main/java/org/zaproxy/zap/session/ScriptBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,668 @@
+import org.parosproxy.paros.model.Session;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.zap.ZAP;
+import org.zaproxy.zap.extension.api.ApiDynamicActionImplementor;
+import org.zaproxy.zap.extension.api.ApiException;
+import org.zaproxy.zap.extension.api.ApiResponse;
+import org.zaproxy.zap.extension.script.ExtensionScript;
+import org.zaproxy.zap.extension.script.ScriptType;
+import org.zaproxy.zap.extension.script.ScriptWrapper;
+import org.zaproxy.zap.extension.sessions.ExtensionSessionManagement;
+import org.zaproxy.zap.extension.sessions.SessionManagementAPI;
+import org.zaproxy.zap.model.Context;
+import org.zaproxy.zap.utils.ApiUtils;
+import org.zaproxy.zap.utils.EncodingUtils;
+import org.zaproxy.zap.view.DynamicFieldsPanel;
+import org.zaproxy.zap.view.LayoutHelper;
+
+/**
+ * The type corresponding to a {@link SessionManagementMethod} for web applications that use uses
+ * Session Management Scripts, which can hopefully model any custom behaviour required
+ */
+public class ScriptBasedSessionManagementMethodType extends SessionManagementMethodType {
+
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT =
+            ExtensionSessionManagement.CONTEXT_CONFIG_SESSION + ".script";
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_NAME =
+            CONTEXT_CONFIG_SESSION_MGMT_SCRIPT + ".name";
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_PARAMS =
+            CONTEXT_CONFIG_SESSION_MGMT_SCRIPT + ".params";
+
+    private static final int METHOD_IDENTIFIER = 2;
+
+    private static final Logger log = Logger.getLogger(ScriptBasedSessionManagementMethod.class);
+
+    /** The Constant METHOD_NAME. */
+    private static final String METHOD_NAME =
+            Constant.messages.getString("sessionmanagement.method.sc.name");
+
+    private static final String API_METHOD_NAME = "scriptBasedSessionManagement";
+
+    public static final String SCRIPT_TYPE_SESSION = "session";
+
+    private static final ImageIcon SCRIPT_ICON_SESSION =
+            new ImageIcon(ZAP.class.getResource("/resource/icon/16/script-session.png"));
+
+    private ExtensionScript extensionScript;
+
+    public class ScriptBasedSessionManagementMethod implements SessionManagementMethod {
+
+        private ScriptWrapper script;
+
+        private Map<String, String> paramValues;
+
+        @Override
+        public boolean isConfigured() {
+            // Always configured
+            return true;
+        }
+
+        @Override
+        public SessionManagementMethodType getType() {
+            return new ScriptBasedSessionManagementMethodType();
+        }
+
+        @Override
+        public WebSession extractWebSession(HttpMessage msg) {
+            SessionScript sessionScript = getScriptInterface(script);
+            if (sessionScript != null) {
+                return sessionScript.extractWebSession(msg);
```

Comment:
Should catch `Exception` to prevent script errors from causing issues elsewhere (same in the other script calls).

Reply:
This was not yet addressed.